pyquil/api/_client.py

```diff
@@ -0,0 +1,241 @@
+import re
+from contextlib import contextmanager
+from datetime import datetime
+from json.decoder import JSONDecodeError
+from typing import Optional, Any, Callable, Iterator, cast
+
+import httpx
+import rpcq
+from dateutil.parser import parse as parsedate
+from dateutil.tz import tzutc
+from qcs_api_client.client import QCSClientConfiguration, build_sync_client
+from qcs_api_client.models import (
+    EngagementWithCredentials,
+    CreateEngagementRequest,
+    EngagementCredentials,
+)
+from qcs_api_client.operations.sync import create_engagement
+from qcs_api_client.types import Response
+from rpcq import ClientAuthConfig
+
+from pyquil.api._errors import ApiError, UnknownApiError, TooManyQubitsError, error_mapping
+from pyquil.api._logger import logger
+
+
+class Client:
+    """
+    Class for housing application configuration and interacting with network resources.
+    """
+
+    _http: Optional[httpx.Client]
+    _config: QCSClientConfiguration
+    _has_default_config: bool
+    _engagement: Optional[EngagementWithCredentials] = None
```

Comment:
I know this is a separate issue, but since we're here, my 10 cents would be to consider:

* Using a raw QCS API client, or possibly a factory.
* A separate compiler client (that also accepts a Pyquil configuration), or possibly a factory.
* An `EngagementManager` class that accepts a QCS API client.